### PR TITLE
frost(sdpa): Rubin MXFP8 prefill closes the gap to FP8 -- row-sum in the MMA (Sigma), zero-copy SF bind, scheduler claims + causal ranking on the quantized d128/d192 flavors

### PR DESCRIPTION
**Pull request by RomanAnders90**
## Why

On Rubin (SM107) the FROST MXFP8 SDPA prefill ran ~10 % behind its per-tensor FP8
sibling on the same shape, while the closed backend's MXFP8 penalty is ~1 %.
Kernel-level profiling attributed the whole deficit to the softmax warps'
register-file row-sum: 63 packed adds per row per KV iteration on the exp'd P
values.  Their presence makes the compiler sink the alpha/stats publish and the
first P-chunk publish to the end of the exp burst, so the MMA warp's BMM2 waits
on the softmax warps every iteration (the FP8 kernel, whose P publish issues
early, does not).  Deleting the row-sum (numerics aside) gave exact parity with
FP8 -- so the fix moves the row-sum out of the register file.

### 1. Sigma: the row-sum rides the tensor core (d128 and d192x128 MXFP8, SM107)

After each Q sub-tile's BMM2 (P x V), the MMA warp issues one N=16 MMA of the
same e4m3 P operand against an all-ones SMEM tile (8 rows x 128 B per CTA at
cga2, filled once at kernel start), accumulating the per-row sum of the
*quantized* P into 16 TMEM columns right after that sub-tile's O.  Plain
`kind::f8f6f4` (P's block scale is 1.0, so the result is identical to a
block-scaled MMA); the same `k_dim`/`TILE_K_HW` pairing as BMM2; the same
`accumulate` predicate, so the first Sigma of a tile overwrites residue.

- Correction warps rescale O + Sigma (span 144 columns at d_v = 128) with the
  same alpha, and the epilogue reads Sigma column 0 as `total_sum`.  The
  softmax warps no longer accumulate a row sum; the stats word that carried it
  is dead and documented as such.
- New barrier `mb_stats_read` (LEADER producer, 128 correction lanes x CTA_MMA
  arrivals, pre-armed): the MMA warp does not overwrite Sigma for the next tile
  until the correction warps have consumed it.  Drained before
  `mb_tmem_dealloc.wait` (P15).  Barrier table and lane arithmetic are in the
  kernel comment block.
- TMEM: d192 map now ends exactly at the 576-column Rubin allocation, d128 at
  564 (import-time raise on overflow + a layout pin test).  The ones tile is
  the last SMEM slab; a module-level guard keeps it under the 256 KiB version-0
  descriptor window (202 KiB at d192, 167 KiB at d128).
- Missing `fence_proxy` after the constant-1.0 P scale-factor fill is now
  covered by the fence after the ones fill.
- The two per-tensor FP8 SM107 kernels gain the same `mb_stats_read` drain
  before TMEM dealloc (they already had the ring; the drain was missing).

**Numerics change by design**: the softmax denominator is now the sum of the
*quantized* P (the FP8 kernels' definition), not of the fp32 exp values.
Measured vs the same fp64 reference: dense max|dO| 1.2e-4 / LSE 2.2e-3, causal
2.3e-2 / 2.6e-2 at d192 S=8K.  No tolerance was widened; the MXFP8 suite is
green on Rubin (same 10 pre-existing e5m2 d128 causal failures as `develop`).

### 2. Zero-copy scale-factor bind (fwd + bwd)

The MXFP8 adapters called `.contiguous()` on the permuted F8_128x4 scale-factor
views the graph path hands them.  For a permuted view that is a copy (three
extra kernels per execute, ~2 % at DSv3 8K) **and** it reorders the bytes of an
opaque swizzled layout.  Both binders now bind the SF tensor's *storage* bytes
(permute to descending-stride order before `.view(torch.int8)`) -- zero-copy
for any dense view, `ValueError` naming the tensor for a non-dense one.  Eight
CPU tests (`test_sdpa_mxfp8_sf_bind.py`).

### 3. Scheduler policy: LPT_L2 reaches the quantized d128 / d192x128 flavors

`SCHED_LPT_L2`'s decode needs `qh_per_kh` / `seqlen_kv` at every decode call
site.  The FP8 kernels already thread them; the two MXFP8 kernels now do (all
ten `_dispatch_decode_*` sites, mirroring the FP8 form).  Claims widen per
flavor (`sched_policies_by_d_shape`; the row-wide floor stays NATURAL):

| row | flavor | before | after |
|---|---|---|---|
| `sdpa_fwd_prefill_sm107_fp8` | (192, 128) | {NATURAL, LPT} | {NATURAL, LPT, LPT_L2} |
| `sdpa_fwd_prefill_sm107_fp8` | (128, 128) | {NATURAL} | {NATURAL, LPT, LPT_L2} |
| `sdpa_fwd_prefill_sm107_mxfp8` | (128, 128), (192, 128) | {NATURAL} | {NATURAL, LPT, LPT_L2} |

Validation on Rubin: O and LSE under LPT and LPT_L2 are bit-identical to
NATURAL on every claimed flavor, dense and multi-wave causal, sentinel-filled O
(Rubin e2e tests, 8 FP8 + 4 MXFP8 cases), and the MXFP8 suite -- whose causal
cases now run a remap -- stays green.  The old NATURAL-only tests are inverted,
not deleted.  Tracker updated.

**Causal ranking (Rubin rows only, `heuristics._sched_points`).**  The
L2-budget rule (LPT_L2 first) is right only where K/V heads are shared.
Perf node, kernel-level, causal, vs NATURAL:

| layout | S=2K | 4K | 8K | 16K | 32K |
|---|---:|---:|---:|---:|---:|
| DSv3 d192x128 H128 (no GQA), LPT | +0.2 % | +16 % | -6.5 % | -13 % | -7.9 % |
| DSv3 d192x128 H128 (no GQA), LPT_L2 | **-9.7 %** | - | +0.3 % | - | -1.2 % |
| Llama d128 H64/8 (GQA), LPT | +5.3 % | +5.8 % | +4.2 % | +2.0 % | +1.1 % |
| Llama d128 H64/8 (GQA), LPT_L2 | **+4.2 %** | **+7.6 %** | **+7.0 %** | **+5.9 %** | **+5.4 %** |

So with `h_q == h_kv` the rule now proposes plain LPT while the grid is at most
24 waves of 2-CTA clusters and NATURAL beyond; GQA graphs keep LPT_L2 first;
every domain member stays an autotune runner; SM100 rows are untouched.

### 4. Missing async-proxy fence after the constant P scale-factor fill (d256 / d512 MXFP8)

Same class as the gap closed on d128 / d192: the 0x7F (E8M0 1.0) `sP_SF` fill
is a generic-proxy SMEM store read once by `tcgen05.cp`; `fence_mbarrier_init`
+ `barrier_cta_sync` do not publish it to the async proxy.  One
`fence_proxy("async.shared")` each, before the CTA barrier.  Byte-identical
output; the MXFP8 suite's d256 / d512 cases pass on Rubin.

## Perf (Rubin, SM clock locked at 2376 MHz, B=1, `benchmark/attention_training/benchmark_single_sdpa.py`, CUDA-event timing, 30 iterations)

FROST backend (`cudnn_oss`), upstream `develop` 85aa39bd -> this branch, TFLOP/s.
Run-to-run noise on this harness: within +-1 % at S >= 4K, up to +-7 % at S=2K
(0.1 ms kernels), so 2K deltas below 8 % are not claims.

| model | mask | dtype | 2K | 4K | 8K | 16K | 32K |
|---|---|---|---:|---:|---:|---:|---:|
| DSv3 d192x128 H128 | dense | **mxfp8** | 2327 -> 2543 (+9.3 %) | 3839 -> 4341 (**+13.1 %**) | 4940 -> 5721 (**+15.8 %**) | 5611 -> 6437 (**+14.7 %**) | 5909 -> 6545 (**+10.8 %**) |
| DSv3 | dense | fp8 | +8.4 % (noise) | -1.7 % | +0.4 % | -0.0 % | +0.1 % |
| DSv3 | causal | **mxfp8** | 1159 -> 1132 (-2.3 %) | 2370 -> 2489 (+5.0 %) | 3726 -> 4171 (**+11.9 %**) | 4781 -> 5275 (**+10.3 %**) | 5468 -> 5945 (**+8.7 %**) |
| DSv3 | causal | fp8 | +0.1 % | -0.7 % | +1.9 % | -0.4 % | +1.0 % |
| Llama d128 H64/8 | dense | **mxfp8** | 1183 -> 1355 (+14.5 %) | 2656 -> 2961 (**+11.5 %**) | 3920 -> 4245 (**+8.3 %**) | 4570 -> 4933 (**+7.9 %**) | 4928 -> 5335 (**+8.3 %**) |
| Llama | dense | fp8 | +2.6 % | +1.9 % | -0.3 % | -0.7 % | -0.4 % |
| Llama | causal | **mxfp8** | 576 -> 629 (+9.2 %) | 1496 -> 1601 (**+7.0 %**) | 2757 -> 3003 (**+8.9 %**) | 3810 -> 4190 (**+10.0 %**) | 4466 -> 4905 (**+9.8 %**) |
| Llama | causal | **fp8** (d128 now runs LPT_L2) | -4.6 % (noise) | +2.2 % | 3064 -> 3241 (**+5.8 %**) | 4237 -> 4458 (**+5.2 %**) | 4917 -> 5174 (**+5.2 %**) |

MXFP8 / FP8 ratio on the FROST backend, before -> after:

| model | mask | 2K | 4K | 8K | 16K | 32K |
|---|---|---:|---:|---:|---:|---:|
| DSv3 | dense | 1.02 -> 1.03 | 0.88 -> 1.01 | 0.87 -> 1.00 | 0.87 -> 1.00 | 0.90 -> 0.99 |
| DSv3 | causal | 0.96 -> 0.94 | 0.89 -> 0.94 | 0.87 -> 0.95 | 0.86 -> 0.95 | 0.89 -> 0.96 |
| Llama | dense | 0.86 -> 0.96 | 0.89 -> 0.97 | 0.90 -> 0.97 | 0.90 -> 0.98 | 0.91 -> 0.98 |
| Llama | causal | 0.86 -> 0.98 | 0.88 -> 0.93 | 0.90 -> 0.93 | 0.90 -> 0.94 | 0.91 -> 0.95 |

For reference the cuDNN backend's MXFP8/FP8 ratio on the same machine is 0.98-1.00 dense.

Kernel-level A/B, PR MXFP8 kernel vs upstream MXFP8 kernel (same machine, same
inputs, standalone adapter, NATURAL, S=8192, B=1, A/B/A 3 x 50 iterations with
a control pair):

| shape | upstream TFLOP/s | PR TFLOP/s | delta | control pair |
|---|---:|---:|---:|---:|
| DSv3 d192x128 H128 dense | 5367 | 5938 | **+10.6 %** | -2.5 % (clock droop across rounds) |
| DSv3 d192x128 H128 causal | 4314 | 4546 | **+5.4 %** | 0.0 % |
| Llama d128 H64/8 dense | 4402 | 4717 | **+7.2 %** | +0.1 % |
| Llama d128 H64/8 causal | 3274 | 3506 | **+7.1 %** | +0.1 % |

Numerics of the PR kernel vs the fp64 reference on those inputs: dense O cos
0.999998, max|dO| 1.2e-4, LSE 2.2e-3; causal O cos 0.999986, max|dO| 2.3e-2,
LSE 2.6e-2 -- the quantized-P denominator.

The same A/B on the per-tensor FP8 kernels (which change only by the barrier
drain) is neutral -- -0.0 / -0.0 / -0.3 / -0.4 % against control pairs of
+-0.2 %, output bit-identical.  Kernel-level MXFP8-vs-FP8 gap after this PR
(same node, same shapes): DSv3 dense 5938 vs 5960 TF/s (0.4 %), causal 4546
vs 4803 (5.4 %); Llama dense 4717 vs 4873 (3.2 %), causal 3506 vs 3591
(2.4 %).  Before: DSv3 dense 5367 vs 5960 (10 %).

## Follow-ups (not in this PR)

- **Sigma on d256 / d512 MXFP8** (static fit analysis done, notes available on request): both fit resource-wise
  (d256: TMEM 512-528 free for Sigma, ones tile at 166/198 KiB, two `mma_ts`
  sites, no new barrier since stats keep a dedicated column, ~+250-300 lines;
  d512: sg1 TMEM 512-528, +1 KiB SMEM under `DESC_VERSION=1`, one `mma_ss`
  site, no new barrier, ~+200-250 lines).  But the d192 root cause does NOT
  transfer: at d256 / d512 the per-tensor FP8 siblings do the same register
  row-sum, so there is no FP8-vs-MXFP8 asymmetry from it and the MXFP8 gap
  there has another cause.  Gate any port on a timing-only "row-sum deleted"
  ceiling on the perf node (Qwen d256 H32/2 cga1; DSv4 d512 H64/1 + H128/1
  cga2, dense + causal); d512 also needs a test module first (none exists).
- `prefill_d256_mxfp8.py` pins `DESC_VERSION = 0` while the config admits
  `STAGES_KV` 2..4: at depth >= 3 with half-precision O the SF tiles start at
  256 KiB and the version-0 descriptor wraps.  Derive it as the f16 sibling
  does (and settle why version 1 broke the d128/d256 MXFP8 tiles before).
- The five SM100 kernels with the same `mb_stats_read` ring have no drain before
  TMEM dealloc; port the P15 drain.
- `config_sm107._validate_smem` does not count the SF slabs or the ones tile
  (~8 KiB; tightest MXFP8 config sits at ~232 of 320 KiB usable).
- LPT_L2 on f16 / d256 / d512 needs the same decode-input threading.
- `frost/compiled_cache.py` is very expensive when `~/.cache` and the `cudnn`
  package live on NFS: one cache-miss `build_plans` on the aarch64 perf node
  spent ~210 s in `scandir`/`stat` pruning the cache root and ~50 s hashing
  every `.py` for the environment manifest (a hit still pays the digest);
  the kernel compile itself is a few seconds.  Worth a cheaper manifest
  (git tree hash / mtime) and a size-accounted prune.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Performance**
  - Improved FP8 and MXFP8 scheduling on Rubin hardware for additional workload shapes, including causal and grouped-query attention.

- **Bug Fixes**
  - Improved MXFP8 scale-factor handling for permuted layouts without unnecessary copying.
  - Fixed synchronization issues affecting FP8 and MXFP8 prefill reliability.
  - Improved log-sum-exp accuracy when statistics are requested for supported FP8 and MXFP8 prefill workloads.

- **Validation**
  - Expanded coverage for numerical consistency, scheduling behavior, memory safety, and scale-factor handling.

### CODE REVIEW

**Comment by RomanAnders90** [comment]
@cudnn-ci-bot run frost

**Review by vedaanta — APPROVED** [review]

**Comment by RomanAnders90** [comment]
@cudnn-ci-bot run frost

File: python/cudnn/sdpa/fwd/kernels/sm107/prefill_d128_fp8.py [hunk 1]
```diff
@@ -1219,6 +1223,27 @@ def _f16_exp_chunk(chunk_S, n: cutlass.Constexpr[int] = 64):
...
+    elems = [chunk_S[i] for i in range(n)]
+    pairs = [fp32_to_fp16(elems[2 * i], elems[2 * i + 1]) for i in range(n // 2)]
+    p_pairs = [ex2_f16x2(w) for w in pairs]
+    words = [f16x2x2_to_fp8_word(p_pairs[2 * g], p_pairs[2 * g + 1], _FP8_TAG_P) for g in range(n // 4)]
+    p_f32 = []
+    for w in p_pairs:
+        lo, hi = f16x2_to_f32(w)
+        p_f32.append(lo)
+        p_f32.append(hi)
+    p_sum = row_reduction_pair(cutlass.Vector.from_elements(tuple(p_f32), cutlass.Float32))
```
[the remark this answers is not in the document]
Comment (RomanAnders90) [reply python/cudnn/sdpa/fwd/kernels/sm107/prefill_d128_fp8.py]: Addressed in 89b1094f. Measured first: summing the f16 P (unpacked to fp32) left the HALF arm's Stats rms 3.6e-4 off the exact log-sum-exp on the dev node, so the f16 exp-argument rounding and MUFU EX2.F16x2's error do not average out. `_f16_exp_chunk_sum` now takes the Stats denominator from a separate fp32 `exp2` of the same biased arguments while O keeps the f16x2 P and the Sigma path. HALF + Stats therefore spends the MUFU the arm saves (honored, not degraded; a stats-less HALF graph is unchanged). `test_fp8_stats_is_the_exact_softmax_lse` is parametrized over the f32 and f16x2 arms under the same 1e-4 bound (f16x2 on d128, where it lives): 10 passed on Rubin.

File: test/python/sdpa/frost/test_sdpa_fwd_dsl_sm107.py [hunk 2]
```diff
@@ -727,3 +786,223 @@ def test_thd_stats_padded_is_appended_to_the_public_signature():
...
+    b, hq, hkv, s = 2, 8, 2, 1024
+    dev = "cuda"
+    qf = torch.randn(b, hq, s, d_qk, device=dev) * 0.5
+    kf = torch.randn(b, hkv, s, d_qk, device=dev) * 0.5
+    vf = torch.randn(b, hkv, s, d_v, device=dev) * 0.5
+
+    def mx(x, h, d, columnwise):
+        data_d, ref_d, swz_d, data_s, ref_s, swz_s = quantize_to_mxfp8(x.contiguous(), b, h, s, d, 32, torch.float8_e4m3fn, with_ref=True)
+        data, ref, swz = (data_s, ref_s, swz_s) if columnwise else (data_d, ref_d, swz_d)
+        return data.permute(0, 2, 1, 3).contiguous().transpose(1, 2), ref.float().reshape(b, h, s, d), swz.contiguous()
```
[the remark this answers is not in the document]
Comment (RomanAnders90) [reply test/python/sdpa/frost/test_sdpa_fwd_dsl_sm107.py]: Correct, `sf_*_ref` are the per-element dequant scales. Fixed in 89b1094f as proposed (`data.float() * scale` in the `[b, h, s, d]` layout before the kernel permutation); the mxfp8 exact-LSE cases pass on Rubin with the corrected reference (max |dLSE| under 1e-4 on d128 and d192x128, dense and causal).

**Comment by RomanAnders90** [comment]
@cudnn-ci-bot run frost
